### DISCUSSION TITLE
Using the Zig build system to ease Windows compilation.

### DISCUSSION THREAD

**Discussion by arrufat** [davisking/dlib#2838] [Ideas]
As mentioned in https://github.com/davisking/dlib/issues/2652#issuecomment-1645640974, creating a `build.zig` file to compile dlib would help reduce the number of bug reports and issues open due to people not being able to install dlib.

I have already started and you can follow the progress [here](https://github.com/arrufat/dlib/tree/zig-build).

It's still very raw, but it can already compile dlib and this simple example:

```c++
#include <dlib/image_io.h>
using rgb_image = dlib::matrix<dlib::rgb_pixel>;
int main(int argc, char** argv) {
    if (argc < 2) return 1;
    rgb_image image;
    dlib::load_image(image, argv[1]);
    dlib::save_jpeg(image, "image.jpg");
    dlib::save_png(image, "image.png");
}
```

All you need to do is add this at the end of the `build.zig` file:

```zig
// this function is already in the build.zig file
pub fn build(b: *std.build.Builder) void {

//  rest of the body

    const exe = b.addExecutable(.{
        .name = "image",
        .target = target,
        .optimize = optimize,
    });
    exe.addCSourceFile("main.cpp", &.{
        "-DDLIB_PNG_SUPPORT=1",
        "-DDLIB_JPEG_SUPPORT=1",
    });
    exe.linkLibrary(dlib);
    b.installArtifact(exe);
}
```

The whole steps look like

1. Download the latest Zig nightly from https://ziglang.org/download/
2. Decompress the file somewhere
3. Use my dlib branch: https://github.com/arrufat/dlib/tree/zig-build
4. Type: `path/to/the/binary/zig build`
5. Done!

If you added those lines to the `build.zig` file, you'll have an `image` executable in `zig-out/bin/` that will read an image from the first argument and save a PNG and a JPEG version.

I've also seen that, on Windows, Zig can target MSVC: https://github.com/adamserafini/zaml/blob/main/builder.py#L21.

So, it should be possible to build the Python module on Windows targeting MSVC.
I will keep developing this when I have free time :)

I haven't managed to use the compiled `libdlib.a` with clang or gcc, apparently I don't know how to pass linker flags properly.
Here's what I am doing, in case someone can spot the mistake:

```sh
g++ -DDLIB_PNG_SUPPORT=1 -DDLIB_JPEG_SUPPORT=1 main.cpp -Izig-out/include -Lzig-out/lib -lz -lpng -ldlib
```

**Comment by arrufat** [comment]
Good news: I know how to pass linker flags properly.
The problem is, since I built with those dependencies with Zig, I need to build the executable with Zig as well:

```
zig c++ -DDLIB_PNG_SUPPORT=1 -DDLIB_JPEG_SUPPORT=1 main.cpp -Izig-out/include -Lzig-out/lib -lz -lpng -ljpeg -ldlib
```

**Comment by pfeatherstone** [comment]
It looks like the Zig build is similar to Meson

**Reply by pfeatherstone** [reply]
If I understand this correctly, Zig ships with several different versions of clang, for cross-compilation, and matching Musl standard libraries ? Does that mean we get to very easily use new language and library features from c++20 and c++23 now ?

**Reply by arrufat** [reply]
Zig only ships with one version of Clang (16 as of today's master build). The next stable version is due on August 3rd, though this doesn't affect us (almost) if we're only using the build system.

But, yes, it ships will the C libraries of plenty of targets, so cross-compiling is trivial, since LLVM can target almost anything. I am using it to generate WebAssembly, it works great.

**Reply by arrufat** [reply]
Also, I think the similarities to Meson are maybe just superficial, since you're using Zig in `build.zig`, which is, of course, Turing complete, contrary to Meson.

**Reply by arrufat** [reply]
You can run `zig targets` and you'll see all the `os`, `abi`, `arch`, `libc`, and `glibc` versions supported. It's mind-blowing, the output is a JSON file of over 21k lines: other the compiler flags are also listed there, such as AVX flags, etc.

**Reply by pfeatherstone** [reply]
Lol. It's laughable that another language provides a better solution to building c++ than a lot of existing c++ specific build tools.

**Reply by pfeatherstone** [reply]
I would be interested to see if it ships with a good up to date c++ standard library. libc++ is typically always behind. I'm pretty sure it still doesn't support special math functions from c++17. So shipping another one which is would be cool. Not that shipping c++ is a priority for Zig.

**Reply by arrufat** [reply]
It's not a priority, but they are always shipping with the latest LLVM, since they develop Zig using it. They're also very active reporting and fixing bugs in LLVM. But yes, I just checked and the mathematical especial functions are not implemented in any Clang version yet.

**Reply by pfeatherstone** [reply]
That's just one thing though. I don't know if coroutines, ranges and other 20 bits are in clang yet.

**Reply by arrufat** [reply]
So, that's why I'd like to try to make a `build.zig` for dlib.
If I manage to build the python module targeting the MSVC ABI, users won't need to install CMake, or Visual Studio, or anything like that on Windows, and we will avoid countless bug reports.

**Reply by arrufat** [reply]
I don't think I'll ever use the `build.zig`, but it's a good way to learn how to use their build system.

**Reply by pfeatherstone** [reply]
Amazing stuff. C++ should definitely have a master build tool that encapsulates compilers for different targets, build system and project configuration stuff. Zig seems to be very close. But I would prefer python as the build language to be honest.

**Reply by pfeatherstone** [reply]
Need to try SCons
